pubtools/_pulp/step.py

```diff
@@ -51,8 +51,19 @@ def new_fn(instance, *args, **kwargs):
 
         return new_fn
 
+    def conditional_skip(self, skip):
+        cond_skip = skip
+        if (
+            self.machine_name == "flush_cdn"
```

Comment:
This isn't the right value.  The step is declared as:

```
@step("Flush CDN cache")
```

That makes the machine name: "flush_cdn_cache"

So this never matches, which is why the branch shows as uncovered.

Overall, while I like the approach of handling it in the decorator, I think it shouldn't be hardcoded for just this case.  My ideal solution would look like this:

```
@step("Flush CDN cache", depends_on=["publish"])
```

The meaning being: this step depends on those steps, so if those steps were skipped, then this one should be too.  If implemented once correctly it could easily be applied to any steps with a similar relationship in the future.

Reply:
i've understand what needs to happen but i'm having a hard time understanding how to have the step decorator accept this additional argument so that it can be used in the StepDecorator class

Reply:
Not tested, but it should be something like this.

```
diff --git a/pubtools/_pulp/step.py b/pubtools/_pulp/step.py
index b606a01..0783ac2 100644
--- a/pubtools/_pulp/step.py
+++ b/pubtools/_pulp/step.py
@@ -9,8 +9,9 @@ LOG = logging.getLogger("pubtools.pulp")
 class StepDecorator(object):
     """Implementation of PulpTask.step decorator. See that method for more info."""
 
-    def __init__(self, name):
+    def __init__(self, name, depends_on):
         self._name = name
+        self._depends_on = depends_on or []
 
     @property
     def human_name(self):
@@ -53,7 +54,10 @@ class StepDecorator(object):
 
     def should_skip(self, instance):
         skip = (getattr(instance.args, "skip", None) or "").split(",")
-        return self.machine_name in skip
+        # We might be *explicitly* skipped,
+        # or *implicitly* skipped if any of our deps are skipped
+        deps_and_self = self._depends_on + [self.machine_name]
+        return any([name in skip for name in deps_and_self])
 
 
 # helpers used in implementation of decorator
diff --git a/pubtools/_pulp/task.py b/pubtools/_pulp/task.py
index 8e7dc92..6e00e74 100644
--- a/pubtools/_pulp/task.py
+++ b/pubtools/_pulp/task.py
@@ -68,7 +68,7 @@ class PulpTask(object):
         return self._args
 
     @classmethod
-    def step(cls, name):
+    def step(cls, name, depends_on=None):
         """A decorator to mark an instance method as a discrete workflow step.
 
         Marking a method as a step has effects:
@@ -94,7 +94,7 @@ class PulpTask(object):
         - The step is considered *failed* if it raised an exception.
         - The step is considered *finished* once all items have been yielded.
         """
-        return StepDecorator(name)
+        return StepDecorator(name, depends_on)
 
     def _basic_args(self):
         # minimum args required for a pulp CLI task
diff --git a/pubtools/_pulp/tasks/common.py b/pubtools/_pulp/tasks/common.py
index 9aed2d8..553c977 100644
--- a/pubtools/_pulp/tasks/common.py
+++ b/pubtools/_pulp/tasks/common.py
@@ -36,7 +36,7 @@ step = PulpTask.step
 class CDNCache(FastPurgeClientService, CdnClientService):
     """Provide features to interact with CDN cache."""
 
-    @step("Flush CDN cache")
+    @step("Flush CDN cache", depends_on=["publish"])
     def flush_cdn(self, repos):
         """Clears the CDN cache for the repositories provided
```

Reply:
thanks, that helps visualize it alot i had not adjusted PulpTask only the StepDecorator class in step.py

Reply:
also i had a much less elegant solution with if conditional inside of should_skip, the way you are showing it here is much cleaner

Reply:
```
        # They should have UD cache flushed.
        to_await.extend(self.flush_ud(repos))
    
        # They should have CDN cache flushed.
        to_await.extend(self.flush_cdn(repos))
    
        # Now make sure we wait for everything to finish.
        for f in to_await:
>           f.result()
E           AttributeError: 'FileRepository' object has no attribute 'result'

pubtools/_pulp/tasks/clear_repo.py:151: AttributeError
```
after some poking around it looks like the tests are telling me clear_repo.py doesn't check if there is no result since cdn cache flush was skipped so it seems like i'll need to do something there to ensure that if skip value is supplied then it shouldn't try to get at that attribute or there is some baseline default result so that any skipped steps would have the attribute still

Reply:
i'm trying to understand:
a skipped function is decorated and is as follows:
```
        def new_fn(instance, *args, **kwargs):
            if self.should_skip(instance):
                LOG.info(
                    "%s: skipped",
                    self.human_name,
                    extra={"event": {"type": "%s-skip" % self.machine_name}},
                )
                return args[0] if args else None
```
and so since flush_cdn is being skipped now based on the skipped publish the return value appears to be None and only the ud flush result is being printed

i remembered as i typed this out about the pdb flag
```
>           f.result()
E           AttributeError: 'FileRepository' object has no attribute 'result'

pubtools/_pulp/tasks/clear_repo.py:151: AttributeError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/lekang/Repos/pubtools-pulp/pubtools/_pulp/tasks/clear_repo.py(151)run()
-> f.result()
(Pdb) print(f)
FileRepository(id='some-filerepo', eng_product_id=123, relative_url='some/publish/url', signing_keys=[], is_sigstore=False, mutable_urls=[])
```
i'm not sure exactly what is happening at that point, so i went to take a look at to_await which is the result of record_push_items which seems to be connected to pushsource i looked at it briefly but i don't think i need to go so deep into that to resolve this

Reply:
It's type confusion about futures vs non-futures again.  The problem is that `flush_cdn` normally returns Futures resolved when the flush has completed, but if the step is skipped then it instead returns `args[0]` which is just a list of repositories rather than futures.

Here's a solution I'd consider for this.  In summary, when defining a step, also make it possible to define what value should be returned if the step is skipped.  In this case it'd make sense to return an empty list (because we did 0 cache flushes in total).

```diff
diff --git a/pubtools/_pulp/step.py b/pubtools/_pulp/step.py
index 3d372d9..7d0f9df 100644
--- a/pubtools/_pulp/step.py
+++ b/pubtools/_pulp/step.py
@@ -4,14 +4,16 @@ import inspect
 from more_executors.futures import f_sequence, f_return
 
 LOG = logging.getLogger("pubtools.pulp")
+UNSET = object()
 
 
 class StepDecorator(object):
     """Implementation of PulpTask.step decorator. See that method for more info."""
 
-    def __init__(self, name, depends_on):
+    def __init__(self, name, depends_on, skipped_value):
         self._name = name
         self._depends_on = depends_on or []
+        self._skipped_value = skipped_value
 
     @property
     def human_name(self):
@@ -29,7 +31,11 @@ class StepDecorator(object):
                     self.human_name,
                     extra={"event": {"type": "%s-skip" % self.machine_name}},
                 )
-                return args[0] if args else None
+                return (
+                    self._skipped_value
+                    if self._skipped_value is not UNSET
+                    else args[0] if args else None
+                )
 
             logger = StepLogger(self)
             args = logger.log_start(args)
diff --git a/pubtools/_pulp/task.py b/pubtools/_pulp/task.py
index 4437ed1..f8f528a 100644
--- a/pubtools/_pulp/task.py
+++ b/pubtools/_pulp/task.py
@@ -4,7 +4,7 @@ from argparse import ArgumentParser, RawDescriptionHelpFormatter
 
 from pubtools.pluggy import task_context
 
-from .step import StepDecorator
+from .step import StepDecorator, UNSET
 
 
 LOG = logging.getLogger("pubtools.pulp")
@@ -68,7 +68,7 @@ class PulpTask(object):
         return self._args
 
     @classmethod
-    def step(cls, name, depends_on=None):
+    def step(cls, name, depends_on=None, skipped_value=UNSET):
         """A decorator to mark an instance method as a discrete workflow step.
 
         Marking a method as a step has effects:
@@ -77,6 +77,8 @@ class PulpTask(object):
         - The method can be skipped if requested by the caller (via --skip argument)
         - Methods that depend on other methods are implicitly skipped if depends_on
             is supplied and that dependant is being skipped
+        - If the method is skipped, it returns either `skipped_value` (if that has
+          been set), or the method's first argument.
 
         Steps may be written as plain blocking functions, as non-blocking
         functions which accept or return Futures, or as generators.
@@ -96,7 +98,7 @@ class PulpTask(object):
         - The step is considered *failed* if it raised an exception.
         - The step is considered *finished* once all items have been yielded.
         """
-        return StepDecorator(name, depends_on)
+        return StepDecorator(name, depends_on, skipped_value)
 
     def _basic_args(self):
         # minimum args required for a pulp CLI task
diff --git a/pubtools/_pulp/tasks/common.py b/pubtools/_pulp/tasks/common.py
index 553c977..cc7d235 100644
--- a/pubtools/_pulp/tasks/common.py
+++ b/pubtools/_pulp/tasks/common.py
@@ -36,7 +36,7 @@ step = PulpTask.step
 class CDNCache(FastPurgeClientService, CdnClientService):
     """Provide features to interact with CDN cache."""
 
-    @step("Flush CDN cache", depends_on=["publish"])
+    @step("Flush CDN cache", depends_on=["publish"], skipped_value=[])
     def flush_cdn(self, repos):
         """Clears the CDN cache for the repositories provided
```

Reply:
that makes alot of sense now, thank you

- flush_cdn accepts a list of repository as the argument
- flush_cdn gets implicitly skipped
- args[0] is this list
- without the above solution then the return value is unexpected when normally flush_cdn returns a list of futures
- when we try to access the result of the future via f.result it is like trying to get at method result from the futures but we are doing so on the repository which it tries to do anyways but result is not an attribute of that object so it complains

it was confusing at first but i think the gist of it has been mostly cleared up

Reply:
It looks like there is a loss of coverage by the change indirectly https://app.codecov.io/gh/release-engineering/pubtools-pulp/pull/266/blob/pubtools/_pulp/services/fakepulp.py but i'm not sure how my changes might have affected that coverage unless i pulled something in by mistake

Reply:
@lekang9 for the purpose of this PR you can just ignore that.  There are two codecov checks, codecov/project and codecov/patch.  codecov/patch is at 100% for your change which means *you* didn't pull down the coverage.

Coverage was somehow decreased during a recent dependency update, it shouldn't block this PR.